src/charm.py

```diff
@@ -0,0 +1,319 @@
+            event.defer()
+            return
+        if not self._default_database_info:
+            self.unit.status = WaitingStatus("Waiting for default database info to be available")
+            event.defer()
+            return
+        if not self._nrf_data_is_available:
+            self.unit.status = WaitingStatus("Waiting for NRF data to be available")
+            event.defer()
+            return
+        content = self._render_config_file(
+            ngapp_port=NGAPP_PORT,
+            sctp_grpc_port=SCTP_GRPC_PORT,
+            sbi_port=SBI_PORT,
+            nrf_url=self._nrf_requires.nrf_url,
+            amf_url=self._amf_hostname,
+            default_database_name=DEFAULT_DATABASE_NAME,
+            amf_database_name=AMF_DATABASE_NAME,
+            database_url=self._default_database_info["uris"].split(",")[0],
+        )
+        if not self._config_file_is_pushed or not self._config_file_content_matches(
+            content=content
+        ):
+            self._push_config_file(
+                content=content,
+            )
+        self._amf_container.add_layer("amf", self._amf_pebble_layer, combine=True)
+        self.unit.status = ActiveStatus()
+
+    def _render_config_file(
+        self,
+        default_database_name: str,
+        amf_database_name: str,
+        amf_url: str,
+        ngapp_port: int,
+        sctp_grpc_port: int,
+        sbi_port: int,
+        nrf_url: str,
+        database_url: str,
+    ):
+        """Renders the AMF config file.
+
+        Args:
+            default_database_name (str): Name of the default (free5gc) database.
+            amf_database_name (str): Name of the AMF database.
+            amf_url (str): URL of the AMF.
+            ngapp_port (int): AMF NGAP port.
+            sctp_grpc_port (int): AMF SCTP port.
+            sbi_port (int): AMF SBi port.
+            database_url (str): URL of the default (free5gc) database.
+            nrf_url (str): URL of the NRF.
+        """
+        jinja2_environment = Environment(loader=FileSystemLoader(CONFIG_TEMPLATE_DIR_PATH))
+        template = jinja2_environment.get_template(CONFIG_TEMPLATE_NAME)
+        content = template.render(
+            ngapp_port=ngapp_port,
+            sctp_grpc_port=sctp_grpc_port,
+            sbi_port=sbi_port,
+            nrf_url=nrf_url,
+            amf_url=amf_url,
+            default_database_name=default_database_name,
+            amf_database_name=amf_database_name,
+            database_url=database_url,
+        )
+        return content
+
+    def _push_config_file(
+        self,
+        content: str,
+    ) -> None:
+        """Writes the AMF config file and pushes it to the container.
+
+        Args:
+            content (str): Content of the config file.
+        """
+        self._amf_container.push(
+            path=f"{CONFIG_DIR_PATH}/{CONFIG_FILE_NAME}",
+            source=content,
+            make_dirs=True,
```

Comment:
We shouldn't have to make the dirs, we're creating the persistent storage in this dir.

Reply:
We may need to validate that this dir exists though before writing the file. i.e. that the storage was attached. Example [here](https://github.com/canonical/sdcore-upf-operator/blob/main/src/charm.py#L239)

Reply:
I agree, as we discuss daily.